### ISSUE TITLE
Support PUT / PATCH / DELETE methods in form

### ISSUE THREAD

**Issue by erdlet** [eclipse-ee4j/krazo#41]
To be able to follow the REST principles in an MVC application it would be great to have a kind of interceptor which converts POST requests to e.g. PUT by some hidden form field's value.

I thought about something like this (currently some pseudo-code, I'm working on a JAX-RS based solution as my first approach runs only on a simple Filter):

```
@PreMatching
public class FormMethodFilter implements ContainerRequestFilter {

  @Override
  public void filter(ContainerRequestContext requestContext) throws IOException {
    if (isFormData(requestContext)) {

      // use EntityStream to receive form data
      String formData = getFormDataFromRequest(requestContext)

      // get hidden field value (or null if not set)
      String hiddenMethod = getHiddenMethod(formData)
     
     // replace current method with hidden one
     if (hiddenMethod != null) {
        requestContext.setMethod(hiddenMethod)
      }
    }
  }
}
```

**Comment by erdlet** [comment]
I added my current implementation to my fork. Unfortunately, when I access the EntityStream of the `ContainerRequestContext`, any form parameter value is null in the controller. When I test this filter with a hard-coded method (and skipping the EntityStream processing) anything works. Seems there is some behavior regarding this EntityStream which I'm not aware of.

**Comment by chkal** [comment]
Hey @erdlet,

reading from the entity stream is problematic in some cases, because it basically prevents other components to read it again. We have similar issues with the CSRF protection, which also needs to read the form (by consuming the entity stream) which breaks `@FormParam` in Wildfly.

You could have a look at #43 for details. @gtudan is currently working on an SPI which should fix these issues.

**Comment by erdlet** [comment]
Hey @chkal,

thanks for the hint. I just had a look on the PR for #43 and I guess it doesn't make sense to
continue on the JAX-RS based form filter until the CSRF issue is solved. 

I'll watch the #43 progress, so I can implement the form method filter with the SPI when it's finished.

**Comment by chkal** [comment]
@erdlet Ok, great! :+1:

**Comment by chkal** [comment]
Fixed via #45
